### PR TITLE
Add spatial slice parallelization tool

### PR DESCRIPTION
**Pull request by keflavich**
This depends on #506.  It splits out the parallelization stuff, and the memory mapped application, into spatial and spectral components

### CODE REVIEW

File: spectral_cube/spectral_cube.py [hunk 1]
```diff
@@ -2823,8 +2830,8 @@ def apply_function_parallel_spectral(self,
             else:
                 pbu = object
 
-            for arg in spectra:
-                _apply_function(arg, outcube, function, **kwargs)
+            for arg in iteration_data:
+                applicator(arg, outcube, function, **kwargs)
                 pbu()
```
Comment (e-koch) [inline spectral_cube/spectral_cube.py]: `LazyComparisonMask` on `outcube` here?

File: spectral_cube/spectral_cube.py [hunk 2]
```diff
@@ -2768,23 +2789,10 @@ def apply_function_parallel_spectral(self,
...
-                    self.mask.include(view=(slice(None), jj, ii)),
-                    ii, jj,
-                   )
-                   for jj in range(shape[1])
-                   for ii in range(shape[2]))
 
         if use_memmap:
             ntf = tempfile.NamedTemporaryFile()
-            outcube = np.memmap(ntf, mode='w+', shape=shape, dtype=np.float)
+            outcube = np.memmap(ntf, mode='w+', shape=self.shape, dtype=np.float)
```
[the remark this answers is not in the document]
Comment (keflavich) [reply spectral_cube/spectral_cube.py]: I think the `reproject` case needs to be its own independent thing; we'll put the parallelization in the reproject package.

File: spectral_cube/spectral_cube.py (hunk 1 above)
Comment (keflavich) [reply spectral_cube/spectral_cube.py]: What do you have in mind?  Should we just be using a `LazyMask` that checks for finite values in the output cube?  Right now we're leaving it untouched, which isn't crazy.

**Comment by keflavich** [comment]
I don't have a clear plan for performance tests.

What are you thinking about in terms of modifying the mask?  I don't think we necessarily need any way to update the mask on the fly within the parallelized loop.  We can just update it after the fact.  No?

File: spectral_cube/spectral_cube.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (e-koch) [reply spectral_cube/spectral_cube.py]: Sounds good

**Comment by keflavich** [comment]
By default, I don't think we should do anything to the mask, then.  We should leave it to the user (or the specific operation) to update the mask after running `apply_function_parallel_spatial`.  At the moment, we don't have any functions using `apply_function_parallel_spatial`, but we should come up with some so we have some practical use and test cases in hand.

**Comment by keflavich** [comment]
@e-koch the smooth tests now actually test this.  I think it's good to merge?

**Comment by e-koch** [comment]
LGTM!

**Pull request merged**